VINDRIKTNING PM1006 active control of measurement requests and fan by ESP

jenicek001 opened the pull request:
## Description:
Particle sensor PM1006 used in IKEA VINDRINKTNING product requires an active control to get measurement results. The original implementation of VINDRIKTNING support in Tasmota was relying the original VINDRIKTNING MCU/PCB to generate measurement requests to PM1006 sensor and controlling the fan.
This functionality is now implemented in Tasmota, so ESP8266 or ESP32 fully control the PM1006 - specifically sensing serial Tx requests once per 60 seconds and switching on the fan 10s before a measurement is taken.

**Related issue (if applicable):** fixes # none

## Checklist:
  - [X] The pull request is done against the latest development branch
  - [X] Only relevant files were touched
  - [X] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [X] The code change is tested and works with Tasmota core ESP32 V.2.0.3
  - [X] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

jenicek001 committed:
VINDRIKTNING PM1006 measurement requests and fan control implemented

This version adds support for ESP8266 or ESP32 to autonomously control VINDRIKTNING PM1006 (without need of original VINDRIKTNING MCU or PCB) - by sending measurement requests every 60s and switching fan 10s before every measurement is taking place.

jenicek001 committed:
Merge branch 'development' of https://github.com/jenicek001/Tasmota into development

arendst commented:
- Move template gpios to correct location. Read file header.
- Learn to use teleperiod. Do NOT program a fixed 60 second interval

jenicek001 committed:
Update tasmota_template.h

New GPIOs moved to correct position

jenicek001 committed:
Update tasmota_template.h

Fixed order of new user selectable GPIO pins in kSensorNames[] and in kSensorNames[]

jenicek001 committed:
Added support for teleperiod

Added support for teleperiod - for VINDRIKTNING fan control. PM1006 values are now actively requested every second.

jenicek001 committed:
Merge branch 'development' into development

jenicek001 commented:
> * Move template gpios to correct location. Read file header.
done
> * Learn to use teleperiod. Do NOT program a fixed 60 second interval
done

jenicek001 committed:
Merge branch 'development' into development

jenicek001 committed:
Merge branch 'development' into development

jenicek001 commented:
Dear maintainers, would somebody please approve my workflow to accept and merge this pull request?

jenicek001 commented:
May I do something to support approval and merging of this pull request?
Thank you and Best Regards,
Jan

Cougar commented on tasmota/include/tasmota_template.h:
```diff
@@ -171,7 +172,7 @@ enum UserSelectablePins {
   GPIO_MCP2515_CS,                     // MCP2515 Chip Select
   GPIO_HRG15_TX, GPIO_HRG15_RX,        // Hydreon RG-15 rain sensor serial interface
-  GPIO_VINDRIKTNING_RX,                // IKEA VINDRIKTNING Serial interface
+  GPIO_VINDRIKTNING_RX,                // IKEA VINDRIKTNING Serial Tx
```
Looks like a typo, shouldn't this be a "Rx"?

jenicek001 replied:
Hello Cougar, thank you for your comments, you are right, typo in comment, fixed.

Cougar commented on tasmota/tasmota_xsns_sensor/xsns_91_vindriktning.ino:
```diff
@@ -96,9 +107,64 @@ bool VindriktningReadData(void) {
+  int i = 0;
+  uint8_t sum = 0;
+
+  if (cmd_len > 2) {
```
In such a simple case where only one command exists, do we need all this code right now?

jenicek001 replied:
As there is just a single command to request PM data from the sensor PM1006, I simplified the code as you propose.

Cougar reviewed (commented):
Some comments

jenicek001 committed:
Update tasmota_template.h

Fixed typo in comment - original pin GPIO_VINDRIKTNING_RX is Rx

jenicek001 committed:
Update xsns_91_vindriktning.ino

As there is just a single command for PM1006 sensor, code is simplified.

jenicek001 committed:
Merge branch 'development' into development

jenicek001 commented:
I just resolved another conflict with several pull requests added later.

Cougar commented on tasmota/tasmota_xsns_sensor/xsns_91_vindriktning.ino:
```diff
@@ -96,9 +107,37 @@ bool VindriktningReadData(void) {
+      }
+    } // if (PinUsed(GPIO_VINDRIKTNING_FAN))
+
+    VindriktningSendRequest(); // request fresh value from VINDRIKTNING PM1006 every second, to get fresh values on web immediately (as values from other sensors)
```
Are you sure it is a good idea to request data every second? I don't know if PM1006 can answer so fast but even then, if it happens after fan has been switched off, the sensor readings might skew until fan has been run again at least one minute.

I would show continuous reading in web only when fan is running all time. Otherwise just mimic VINDRIKTNING behavior and update web reading also only at every teleperiod. What would you think about that?

jenicek001 replied:
Hello Cougar, thank you for reviewing the code.

Original microcontroller in IKEA VINDRIKTNING is requesting PM1006 sensor for a new value each second, Tasmota before my chages use ESP to only read sensor values (which are available each second).
For cases of original IKEA board fully replaced by an ESP board (e.g. LaskaKit ESP-VINDRIKTNING), Tasmota is driving PM1006 sensor itself (this modification) and I decided not to change this original behaviour.

Based on arendst's comment above, I learned purpose of teleperiod and designed fan control to provide enough time advance before measurement (and values export) takes place - for fresh air to be delivered into PM1006 sensor to provide accurate results.
Yes, you are right, web UI values with stopped fan may not be accurate for longer teleperiods, so it makes sense to stop updating them. I can implement this change, of course.

Additional note: As IKEA VNDRIKTNING device is well suitable to fit also an HT sensor (e.g. SCD40/41 supported by Tasmota), I tuned timing of PM1006 fan switching for different teleperiods (by lot of measurements and comparing reported values to a reference temp meter), because VINDRIKTNING housing is inside heated by own ESP dissipated heat (esp. for ESP32). The tiny PM1006 fan needs ~1 minute to decrease internal temperature to the ambient level. This value is a parameter in source code and could be potentially turned into a runtime parameter.

Regards,
Jan

arendst commented:
unstale

arendst commented:
bump

jenicek001 committed:
Merge branch 'development' into development

jenicek001 committed:
Merge branch 'development' into development

jenicek001 commented:
Hello, is there some blocking point to merge this pull request?

jenicek001 committed:
Merge branch 'development' into development

jenicek001 committed:
Merge branch 'development' into development

arendst commented:
Bump

jenicek001 committed:
Merge branch 'development' into development

jenicek001 committed:
Merge branch 'development' into development

jenicek001 commented:
Hello, is there something I can do to support merge? I successfully use my code on 2 VINDRIKTNINGS for almost 1 year since my pull request, I assume it may be useful to more enthusiasts.
Thank you in advance.

brozikcz commented:
I'm not sure but someone like @s-hadinger can approve or provide some advice on how to merge it.

jenicek001 commented:
Dear colleagues, does it make sense to delete this working code extension? Did I break some merge rules or what is wrong to merge it?
Thank you,
Jan

arendst commented:
What is wrong is the fact that non of my initial remarks have been fixed. In addition this is an edge case for some users willing to canibalize their device.

jenicek001 commented:
Hello Theo,
I read your remarks and just few days after I fixed my code to comply with them: https://github.com/arendst/Tasmota/pull/15878#issuecomment-1176156076
Did I miss something important?
Regards,
Jan

jenicek001 commented:
Hello Theo,
would you please be more specific what do you mean exactly?
> What is wrong is the fact that non of my initial remarks have been fixed. In addition this is an edge case for some users willing to canibalize their device.
I did my best to follow your notes since creation of this pull request one year ago, so what should I fix to match with your expectations?
Best regards,
Jan
